odl/solvers/functional/functional.py

```diff
@@ -884,3 +885,98 @@ def convex_conj(self):
         """
         return self.functional.convex_conj.translated(
             self.linear_term)
+
+
+class FunctionalProduct(Functional, OperatorPointwiseProduct):
+
+    """The ``Functional`` representing ``f(.) * g(.)``."""
+
+    def __init__(self, left, right):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        left, right : `Functional`
+            Functionals in the product.
+        """
+        if not isinstance(left, Functional):
+            raise TypeError('`func1` {} is not a `Functional` instance'
+                            ''.format(left))
+        if not isinstance(right, Functional):
+            raise TypeError('`func2` {} is not a `Functional` instance'
+                            ''.format(right))
+
+        OperatorPointwiseProduct.__init__(self, left, right)
+        Functional.__init__(self, left.domain, linear=False,
+                            grad_lipschitz=np.nan)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional."""
+        func = self
+
+        class FunctionalProductGradient(Operator):
```

Comment:
Either add documentation, or add a new class `OperatorLeftFunctionalMultiplication` that generalizes this.